### PR TITLE
[AIRFLOW-6055] Option for exponential backoff in Sensors

### PR DESCRIPTION
**Pull request by msumit**
Make sure you have checked _all_ steps below.

### Jira

- [ ] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issue https://issues.apache.org/jira/browse/AIRFLOW-6055

### Description

- [ ] Adding an option in sensors to do exponential backoff before poke or schedule again.

### Tests

- [ ] Added a couple of unit tests around the new function

### Commits

- [ ] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### CODE REVIEW

File: tests/sensors/test_base_sensor.py [hunk 1]
```diff
@@ -502,3 +502,28 @@ def test_sensor_with_invalid_timeout(self):
...
+    def test_sensor_with_exponential_backoff_off(self):
+        sensor = self._make_sensor(
+            return_value=None,
+            poke_interval=5,
+            timeout=60,
+            exponential_backoff=False)
+
+        started_at = timezone.utcnow() - timedelta(seconds=10)
+        self.assertEqual(sensor._get_next_poke_interval(started_at, 1), sensor.poke_interval)
+        self.assertEqual(sensor._get_next_poke_interval(started_at, 2), sensor.poke_interval)
```
Comment (ashb) [inline tests/sensors/test_base_sensor.py:515]: I'd rather we had the exact values here in the tests - this feels like it's just running the code again, but not actually testing the behaviour.

File: tests/sensors/test_base_sensor.py [hunk 2]
```diff
@@ -502,3 +502,28 @@ def test_sensor_with_invalid_timeout(self):
...
+            poke_interval=5,
+            timeout=60,
+            exponential_backoff=True)
+
+        started_at = timezone.utcnow() - timedelta(seconds=10)
+
+        interval1 = sensor._get_next_poke_interval(started_at, 1)
+        interval2 = sensor._get_next_poke_interval(started_at, 2)
+
+        self.assertTrue(interval2 >= sensor.poke_interval >= interval1)
```
Comment (ashb) [inline tests/sensors/test_base_sensor.py]: Here too.

**Review by ashb — CHANGES_REQUESTED** [review]

File: tests/sensors/test_base_sensor.py (hunk 1 above)
Comment (msumit) [reply tests/sensors/test_base_sensor.py:515]: `sensor.poke_interval` is an exact value `5`. The test is to make sure that between diff tries, the output is not changing.

File: tests/sensors/test_base_sensor.py (hunk 2 above)
Comment (msumit) [reply tests/sensors/test_base_sensor.py]: problem is that this method calculates sha1, which is based upon the started_at, which is relative to the current time.

File: tests/sensors/test_base_sensor.py (hunk 1 above)
Comment (ashb) [reply tests/sensors/test_base_sensor.py:515]: Oh yes, this one is fine then.

File: tests/sensors/test_base_sensor.py (hunk 2 above)
Comment (ashb) [reply tests/sensors/test_base_sensor.py]: In which case use mocking to make `timezone.utcnow` return a known constant value in the tests.

Or if the "jitter" caused by this is small then 
```
self.assertGreaterThanEqual(interval1, 5)
self.assertLessThanEqual(interval1, 5.5)

self.assertGreaterThanEqual(interval2, 6)
self.assertLessThanEqual(interval2, 7)
```

Those values are not right, but this is the sort of thing I'd like to see.

**Review by kaxil — APPROVED** [review]

File: tests/sensors/test_base_sensor.py (hunk 2 above)
Comment (msumit) [reply tests/sensors/test_base_sensor.py]: mocked timezone

Comment (ashb) [reply tests/sensors/test_base_sensor.py]: You mocked the timezone, but didn't add the extra tests that I asked for. Please could you add them?

**Pull request merged**